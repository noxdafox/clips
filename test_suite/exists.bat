(clear)      ; Test exists
(watch activations)
(watch facts)

(defrule test
   (exists (a ?))
   =>)
(reset)
(assert (a 1))
(assert (a 2))
(assert (a 3))
(retract 1)
(retract 3)
(retract 2)
(assert (a 4))
(clear)

(defrule test
   (a ?x)
   (exists (b ?x ?))
   =>)
(reset)
(assert (b 1 1))
(assert (b 1 2))
(assert (a 1))
(assert (a 2))
(retract 1)
(retract 2)
(assert (b 2 2))
(assert (b 1 2))
(assert (b 2 3))
(assert (b 1 3))
(retract 7)
(retract 5)
(retract 3)
(unwatch facts)
(unwatch activations)
(clear)
(defrule foo (a) (exists (a) (a)) =>)
(reset)
(assert (a))
(agenda)
(retract 1)
(agenda)
(clear) ; 6.3 Issue
(unwatch all)   

(defrule detect
   (b ?b)
   (exists (a))
   (test (eq ?b 2))
   =>
   (printout t "Rule Fired" crlf))
(assert (a))
(assert (b 1))
(run)
(reset)
(assert (b 1))
(assert (a))
(run)
(reset)
(assert (a))
(assert (b 2))
(run)
(reset)
(assert (b 2))
(assert (a))
(run)
(clear) ; 6.3 Issue

(deffacts xy
    (SAD T 1 T07 "12345678"))

(defrule x
   (exists (SAD T ?ix T07 ?str)
       (test (>= (str-length ?str) 0))) 
   (test (=  1 1))
   =>)
(reset)
(agenda)
(clear)
