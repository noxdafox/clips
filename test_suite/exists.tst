(set-strategy depth)
(unwatch all)
; exists.bat test
(clear)
(open "Results//exists.rsl" exists "w")
(dribble-on "Actual//exists.out")
(batch "exists.bat")
(dribble-off)
(load "compline.clp")
(printout exists "exists.bat differences are as follows:" crlf)
(compare-files "Expected//exists.out" "Actual//exists.out" exists)
; close result file
(close exists)
