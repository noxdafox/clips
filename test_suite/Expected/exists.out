TRUE
CLIPS> (batch "exists.bat")
TRUE
CLIPS> (clear)      ; Test exists
CLIPS> (watch activations)
CLIPS> (watch facts)
CLIPS> 
(defrule test
   (exists (a ?))
   =>)
CLIPS> (reset)
<== f-0     (initial-fact)
==> f-0     (initial-fact)
CLIPS> (assert (a 1))
==> f-1     (a 1)
==> Activation 0      test: *
<Fact-1>
CLIPS> (assert (a 2))
==> f-2     (a 2)
<Fact-2>
CLIPS> (assert (a 3))
==> f-3     (a 3)
<Fact-3>
CLIPS> (retract 1)
<== f-1     (a 1)
CLIPS> (retract 3)
<== f-3     (a 3)
CLIPS> (retract 2)
<== f-2     (a 2)
<== Activation 0      test: *
CLIPS> (assert (a 4))
==> f-4     (a 4)
==> Activation 0      test: *
<Fact-4>
CLIPS> (clear)
==> f-0     (initial-fact)
CLIPS> 
(defrule test
   (a ?x)
   (exists (b ?x ?))
   =>)
CLIPS> (reset)
<== f-0     (initial-fact)
==> f-0     (initial-fact)
CLIPS> (assert (b 1 1))
==> f-1     (b 1 1)
<Fact-1>
CLIPS> (assert (b 1 2))
==> f-2     (b 1 2)
<Fact-2>
CLIPS> (assert (a 1))
==> f-3     (a 1)
==> Activation 0      test: f-3,*
<Fact-3>
CLIPS> (assert (a 2))
==> f-4     (a 2)
<Fact-4>
CLIPS> (retract 1)
<== f-1     (b 1 1)
CLIPS> (retract 2)
<== f-2     (b 1 2)
<== Activation 0      test: f-3,*
CLIPS> (assert (b 2 2))
==> f-5     (b 2 2)
==> Activation 0      test: f-4,*
<Fact-5>
CLIPS> (assert (b 1 2))
==> f-6     (b 1 2)
==> Activation 0      test: f-3,*
<Fact-6>
CLIPS> (assert (b 2 3))
==> f-7     (b 2 3)
<Fact-7>
CLIPS> (assert (b 1 3))
==> f-8     (b 1 3)
<Fact-8>
CLIPS> (retract 7)
<== f-7     (b 2 3)
CLIPS> (retract 5)
<== f-5     (b 2 2)
<== Activation 0      test: f-4,*
CLIPS> (retract 3)
<== f-3     (a 1)
<== Activation 0      test: f-3,*
CLIPS> (unwatch facts)
CLIPS> (unwatch activations)
CLIPS> (clear)
CLIPS> (defrule foo (a) (exists (a) (a)) =>)
CLIPS> (reset)
CLIPS> (assert (a))
<Fact-1>
CLIPS> (agenda)
0      foo: f-1,*
For a total of 1 activation.
CLIPS> (retract 1)
CLIPS> (agenda)
CLIPS> (clear) ; 6.3 Issue
CLIPS> (unwatch all)   
CLIPS> 
(defrule detect
   (b ?b)
   (exists (a))
   (test (eq ?b 2))
   =>
   (printout t "Rule Fired" crlf))
CLIPS> (assert (a))
<Fact-1>
CLIPS> (assert (b 1))
<Fact-2>
CLIPS> (run)
CLIPS> (reset)
CLIPS> (assert (b 1))
<Fact-1>
CLIPS> (assert (a))
<Fact-2>
CLIPS> (run)
CLIPS> (reset)
CLIPS> (assert (a))
<Fact-1>
CLIPS> (assert (b 2))
<Fact-2>
CLIPS> (run)
Rule Fired
CLIPS> (reset)
CLIPS> (assert (b 2))
<Fact-1>
CLIPS> (assert (a))
<Fact-2>
CLIPS> (run)
Rule Fired
CLIPS> (clear) ; 6.3 Issue
CLIPS> 
(deffacts xy
    (SAD T 1 T07 "12345678"))
CLIPS> 
(defrule x
   (exists (SAD T ?ix T07 ?str)
       (test (>= (str-length ?str) 0))) 
   (test (=  1 1))
   =>)
CLIPS> (reset)
CLIPS> (agenda)
0      x: *
For a total of 1 activation.
CLIPS> (clear)
CLIPS> (dribble-off)
